MP Health - local/file based health checks. Introduce the _create interval_.

Channyboy opened the pull request:
- [ ] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

################################################################################################

Channyboy committed:
Introduce the FileCreateInterval config.

Channyboy committed:
Update behaviour to create all 3 health files when all three report UP statuses

Channyboy committed:
updates to FATs in io.openliberty.microprofile.health.file.healthcheck.internal_fat; Use create interval of 1s (to grant the machiens some reprieve)

Channyboy committed:
Update instantOn FAT test

Channyboy commented:
#build

Channyboy commented:
#build #libby ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6f92679b-094f-4270-8eec-50cb34850870) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=eb294e67-8c81-4745-848b-13ff94fd63ec) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy commented:
#build #libby

Channyboy commented:
#build #libby ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=d4e8455a-3d24-44bb-b74d-deb6fc035447) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=71b5f8d4-dc52-43e4-b794-7bc6693fa64d) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy committed:
Variety of changes:

- Allow for ISE; TimerTask is already when application shuts down resulting in component metadata being missing/null.
- Allow for CWMMH0050E. Similar reason to above, but regarding CDI querying.
- Fix message code.
- lock on trying to read configAdminMap

Channyboy commented:
#build #libby ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=419d3a6c-527f-4223-9e9b-fccb0d35fdf7) - :white_check_mark: completed successfully!) ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=4ba20b2e-0506-45bb-86e4-3342c04849e8) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

is273 commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -19,4 +19,7 @@ mpHealth=MicroProfile Health
+fileUpdateInterval.desc=The value that you specify sets the interval at which the health checks are queried for the file-based health checks. The value is a number followed by an optional time unit of ms for milliseconds or s for seconds. If no time unit is specified, the value is in seconds by default.
+
+fileCreateInterval=Interval of time at which to create the file-based health checks.
+fileCreateInterval.desc=The value that you specify sets the interval at which the health checks are queried until an UP status is resolved and the health check file is created. The value is a number followed by an optional time unit of ms for milliseconds or s for seconds. If no time unit is specified, the value is in seconds by default. The default value is 100ms.
```
"If no time unit is specified, the value is in seconds by default. The default value is 100ms." change to "If no time unit is specified for a value, the value is in seconds by default. If no value is specified, the default value is 100 ms."

is273 commented on dev/io.openliberty.microprofile.health.internal.common/resources/io/openliberty/microprofile/health/resources/Health.nlsprops:
```diff
@@ -79,6 +79,18 @@ file.healthcheck.file.delete.fail.CWMMH0106W=CWMMH0106W: Unable to delete existi
 file.healthcheck.file.delete.fail.CWMMH0106W.explanation=The file-based health checks require the creation and use of the files "live", "ready" and "started" in the /health directory. For the file-based health checks to function correctly, these files must not exist when the runtime starts. The runtime has attempted to delete the existing file was unable to.
 file.healthcheck.file.delete.fail.CWMMH0106W.useraction=The user must identify the root cause preventing the runtime from deleting the files.
 
+file.healthcheck.system.inconsistency.CWMMH0107W=CWMMH0107W: The runtime has detected inconsistencies with the local file-based health check mechanism. The file-based health check function will be disabled.
```
Change to:

file.healthcheck.system.inconsistency.CWMMH0107W=CWMMH0107W: The runtime detected inconsistencies with the local file-based health check mechanism. The file-based health check mechanism is disabled.
file.healthcheck.system.inconsistency.CWMMH0107W.explanation=Inconsistencies were detected with the file system. For example, the "started" file is present before the runtime creates the "started" file.
file.healthcheck.system.inconsistency.CWMMH0107W.useraction=Delete the <server.output.dir>/health directory and restart the server.

is273 commented on dev/io.openliberty.microprofile.health.internal.common/resources/io/openliberty/microprofile/health/resources/Health.nlsprops:
```diff
@@ -79,6 +79,18 @@ file.healthcheck.file.delete.fail.CWMMH0106W=CWMMH0106W: Unable to delete existi
 file.update.interval.config.invalid.CWMMH01010W.explanation=The file update interval configuration only accepts a number followed by an optional time unit of "ms" for milliseconds or "s" for seconds. If no time unit is specified, the value is in seconds by default.
-file.update.interval.config.invalid.CWMMH01010W.useraction=Input a correct value for the file update interval configuration.
\ No newline at end of file
+file.update.interval.config.invalid.CWMMH01010W.useraction=Input a correct value for the file update interval configuration.
```
Change to:
 file.create.interval.config.invalid.CWMMH01011W=CWMMH01011W: An invalid value [{0}] was provided for the file create interval. The value provided must consist of a number followed by an optional time unit of "ms" for milliseconds or "s" for seconds. If no time unit is specified, the value is in milliseconds by default. The runtime defaults to 100 milliseconds.
file.create.interval.config.invalid.CWMMH01011W.explanation=The file create interval configuration accepts only values with a number followed by an optional time unit of "ms" for milliseconds or "s" for seconds. If no time unit is specified, the value is in milliseconds by default.
file.create.interval.config.invalid.CWMMH01011W.useraction=Specify a correct value for the file create interval configuration.

is273 commented on dev/io.openliberty.microprofile.health.internal.common/resources/io/openliberty/microprofile/health/resources/Health.nlsprops:
```diff
@@ -79,6 +79,18 @@ file.healthcheck.file.delete.fail.CWMMH0106W=CWMMH0106W: Unable to delete existi
+file.create.interval.config.invalid.CWMMH01011W.explanation=The file create interval configuration only accepts a number followed by an optional time unit of "ms" for milliseconds or "s" for seconds. If no time unit is specified, the value is in milliseconds by default.
+file.create.interval.config.invalid.CWMMH01011W.useraction=Input a correct value for the file create interval configuration.
+
+file.create.interval.config.only.set.CWMMH01012W=CWMMH01012W: The file create interval configuration was set without the file update interval configuration set. The file update interval is required to be configured to enable the file-based health check.
```
Change to:
file.create.interval.config.only.set.CWMMH01012W=CWMMH01012W: The file create interval configuration was set before the file update interval was configured. The file-based health check is disabled when the file update interval is not configured.
file.create.interval.config.only.set.CWMMH01012W.explanation=Configuring only the file create interval does not enable the file-based health check mechanism. 
file.create.interval.config.only.set.CWMMH01012W.useraction=Configure the file update interval to enable the file-based health check mechanism.

Channyboy committed:
apply ID review comments

is273 reviewed (dismissed):
LGTM

Channyboy commented:
#build #libby ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=8e4e1de9-62ae-457b-bdf5-3b1b842cadf5) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=5e0b385c-9d1d-4733-a674-5455d943d468) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy commented:
rebased
#build #libby ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=cd9c97f4-0f4d-443c-908f-7f95e3ce3085) - :x: completed with errors/failures.) ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e06e10a4-5a1b-4e9e-bb1f-1428739df8f8) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/internal/HealthCheck40ServiceImpl.java:
```diff
@@ -205,57 +259,106 @@ protected void activate(ComponentContext cc, Map<String, Object> properties) {
-                if (isFileHealthCheckingEnabled()) {
-                    startFileHealthCheckProcesses();
+                /*
+                 * If were were already in the update phase when config was modified.
```
typo

pgunapal commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -18,5 +18,8 @@
-fileUpdateInterval.desc=The value that you specify sets the interval at which the health checks are queried for the file-based health checks. The value is a number followed by an optional time unit of ms for milliseconds or s for seconds. If no time unit is specified, the value is in seconds by default.
\ No newline at end of file
+checkInterval=Interval of time at which to update the file-based health checks.
+checkInterval.desc=The value that you specify sets the interval at which the health checks are queried for the file-based health checks. The value is a number followed by an optional time unit of ms for milliseconds or s for seconds. If no time unit is specified, the value is in seconds by default.
```
Is there a default value for `checkInterval`?

Channyboy replied:
There is no default value for `checkInterval` formerly `fileUpdateInterval`. The existence/configuration of it enables functionality.

pgunapal replied:
Might want to state that if its empty value, the functionality is disabled?

pgunapal commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -17,7 +17,8 @@
 
     <OCD id="io.openliberty.microprofile.health" name="%mpHealth" description="%mpHealth.desc" ibm:alias="mpHealth" ibm:beta="true">
-        <AD id="fileUpdateInterval" name="%fileUpdateInterval" description="%fileUpdateInterval.desc" required="false" type="String" ibm:beta="true"/>
+        <AD id="checkInterval" name="%checkInterval" description="%checkInterval.desc" required="false" type="String" ibm:beta="true"/>
```
Would be good to set the `ibm:type="duration(ms)"` here, also the default `default="100s"` value

Channyboy replied:
There is no default value for `checkInterval`. The use of duration with its current localized support could lead to scenarios where someone types in _invalid_ values would be recognized as a time unit of some sort. THe direction is that we want to support the "standard" time units.

pgunapal commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -17,7 +17,8 @@
     <OCD id="io.openliberty.microprofile.health" name="%mpHealth" description="%mpHealth.desc" ibm:alias="mpHealth" ibm:beta="true">
-        <AD id="fileUpdateInterval" name="%fileUpdateInterval" description="%fileUpdateInterval.desc" required="false" type="String" ibm:beta="true"/>
+        <AD id="checkInterval" name="%checkInterval" description="%checkInterval.desc" required="false" type="String" ibm:beta="true"/>
+        <AD id="startupCheckInterval" name="%startupCheckInterval" description="%startupCheckInterval.desc" required="false" type="String" ibm:beta="true"/>
```
Good to set the following here: `ibm:type="duration(ms)" default="100ms"`

Channyboy replied:
Providing a default values prevents us from knowing if the configuration was set from the user (or provided as a default). We need user interaction to better gauge the intent on what the user wanted. With regards to the type, see above comment.

Channyboy commented:
#build #libby
 ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=f199d0fb-64c1-4253-882b-1c309bd4a850) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=5ff01c10-f3d5-4061-8312-716f58d972cd) - :x: completed with errors/failures.)
@is273 we've updated the the config name, and added an extra little blurb about 0, 0s, 0ms and empty values for the `checkInterval.desc` in `metatype.properties`. May you please re-review.

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal reviewed (dismissed):
LGTM

Channyboy committed:
Use the new config values fileUpdateInterval -> checkInterval & fileCreateInterval -> startupCheckInterval

Channyboy commented:
#build #libby ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=cce37653-a9be-4be6-877c-8b6cf40f15f2) - :x: completed with errors/failures.) ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=34da8e11-f8a2-4b9c-a904-0a12e4274747) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=8746d1c8-c03d-48aa-bc47-482b625e29dd) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy committed:
synchronize getBeanManger call

Channyboy commented:
#build #libby ([view Open Liberty Delivery Requirements Verification](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=d228b994-29ca-4c11-a710-1c743f60bc6b) - :white_check_mark: completed successfully!) ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=5e9e1e7b-7d68-40a4-8dff-e63d1edbd827) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/services/impl/HealthCheck40CDIBeanInvokerImpl.java:
```diff
@@ -150,7 +150,8 @@ private Set<Object> getHealthCheckBeans(BeanManager beanManager, Annotation hcQu
     }
 
-    private BeanManager getBeanManager(String appName, String moduleName) {
+    private synchronized BeanManager getBeanManager(String appName, String moduleName) {
```
Would this cause any performance issues for the normal mpHealth functionality, every time we query the endpoints simultaneously?

Channyboy replied:
Possibly if that happens. But the bean manager list is small enough that this should return very quickly. The issue is during initial app deployment, where the map can be modified while being retrieved resulting in a concurrent modification exception.

pgunapal replied:
We did see some cases in the past, where an EAR contained at least 10 different WARs/modules, so the bean list might get bigger, in that case. Would be good to get that scenario tested, during our performance testing.

pgunapal reviewed (approved):
LGTM - need a follow up test for performance to verify normal mpHealth functionality still works when endpoints are hit simultaneously.
